### ISSUE TITLE
Unable to update istio configuration in air-gapped environment in Rancher v2.5.x

### ISSUE THREAD

**Issue by axeal** [rancher/rancher#30842]
**What kind of request is this (question/bug/enhancement/feature request):**

Bug

**Steps to reproduce (least amount of steps as possible):**

1. Provision a Rancher v2.5.5 instance and an air-gapped custom cluster (specifically it should not have access to github.com, I mimicked this in my non-air-gapped lab environment by adding the following to the coredns ConfigMap to prevent the resolution of github.com domains):
```
  github.com:53 {
    errors
    cache 30
    loop
    forward . 192.168.200.1
  }
```
2. Install Monitoring in the cluster: Cluster Explorer -> Apps & Marketplace -> Charts -> Monitoring
3. Install Istio in the cluster: Cluster Explorer -> Apps & Marketplace -> Charts -> Istio
4. Upgrade the Istio installation, changing the configuration, e.g. enable the Egress Gateway component.

**Result:**

Observe the upgrade has no effect and that the logs of the `istioctl-installer` Job Pod show the following error (in the case of mimicking air-gap with coredns ConfigMap change the exact `lookup github.com` error message will differ slightly):

```
Upgrade version check passed: 1.7.3 -> 1.7.3.

2021-01-13T14:30:17.398813Z     info    Error: failed to generate Istio configs from file [/app/istio-base.yaml] for the current version: 1.7.3, error: Get "https://github.com/istio/istio/releases/download/1.7.3/istio-1.7.3-osx.tar.gz": dial tcp: lookup github.com on 10.43.0.10:53: no such host

Error: failed to generate Istio configs from file [/app/istio-base.yaml] for the current version: 1.7.3, error: Get "https://github.com/istio/istio/releases/download/1.7.3/istio-1.7.3-osx.tar.gz": dial tcp: lookup github.com on 10.43.0.10:53: no such host
```

**Other details that may be helpful:**

There is an issue reporting this behaviour in the istio/istio GH repo at https://github.com/istio/istio/issues/28995

Behaviour originates from the `istio upgrade` command in the istioctl-installer Job https://github.com/rancher/istio-installer/blob/master/scripts/create_istio_system.sh

The error appears to come [from here](https://github.com/istio/istio/blob/1.7.3/operator/cmd/mesh/upgrade.go#L140) in the istio codebase, but at a first glance I could not determine the difference between the `istioctl install` and `istioctl upgrade` that means this is encountered only on the upgrade command, and not on the install.

**Environment information**
- Reproduced in Rancher v2.5.5, encountered by user in v2.5.2

gz#13932

**Comment by bashofmann** [comment]
I can also reproduce this issue. Installation works, but upgrade fails. Explicitly specifying a local manifests directory, like I tried in https://github.com/rancher/istio-installer/pull/15 does not help. The problem still persists.

**Comment by brendarearden** [comment]
TLDR; If a current version of istio exists (which is always true for upgrade) then it sets a url for installPackagePath and ignores passed in values. 

I have explained and linked code here https://github.com/istio/istio/issues/28995#issuecomment-772323915 with details of the root cause. This is an upstream bug and not something we are able to resolve with a code fix on our end at this time.

**Comment by brendarearden** [comment]
The upstream repository closed out the original issues without merging in a fix so this is still an issue with the upstream Istio Operator when `Upgrade` is used. I have commented on those issues saying the fix is still needed.

**Comment by xiaoluhong** [comment]
@brendarearden  To fix this temporarily, I add an nginx service to the istio-installer image. add https Server to the NGINX configuration, use self-signed SSL certificates. Then, configure /etc/hosts to point github.com to local 127.0.0.1.   The Dockerfile https://github.com/xiaoluhong/istio-installer/blob/test-1.8.3/Dockerfile

**Comment by brendarearden** [comment]
https://github.com/istio/istio/issues/32689 << opened new issue upstream as I was unable to get the original issue re-opened

**Comment by trieszecg** [comment]
as a workaround is it possible to somehow export HTTPS_PROXY env var for the istioctl-installer?

**Comment by bashofmann** [comment]
I created a POC in istio-installer to serve the needed release artefacts locally from a pod in the cluster. The istio-installer pod patches its `/etc/hosts` file to route traffic to github.com to the internal pod.

istio-installer: https://github.com/bashofmann/istio-installer/tree/airgap-upgrades
helm chart: https://github.com/bashofmann/charts-1/tree/istio-airgap-upgrades/packages/rancher-istio/charts

It's not perfect yet:
* The amount of release artefacts in the image is just the first response page of the github api request. There are likely important versions missing.
* The fake github.com certificates are hardcoded in the images
* It probably won't work with PSPs when the istio-installer container is run as non-root

**Comment by cbron** [comment]
release-note: upgrading will be limited to versions available at time of build, @brendarearden can explain further.

**Comment by brendarearden** [comment]
release-note: PSP is now enabled by default

**Comment by brendarearden** [comment]
For QA: 

The istio PSP is now enabled by default and a new value `installer.releaseMirror.enabled` has been added to the rancher-istio chart to enable and disable the server that supports airgapped upgrades. `installer.releaseMirror.enabled` is set to false by default.

We need to do the standard istio testing with this change, but I also want to call out specific test cases for this issue: 

|Test Case| Expected Outcome|
|------------|----------------|
| Airgapped env: Install, upgrade, uninstall | Install and Uninstall completes successfully, Upgrade Fails|
| Airgapped env: Install, upgrade, uninstalled with `installer.releaseMirror.enabled=true` | Install, Upgrade and Uninstall completes successfully|
| Airgapped env: Install with `installer.releaseMirror.enabled=false` (its false by default so you do not need to set this), Upgrade with `installer.releaseMirror.enabled=true`,  Uninstall  with `installer.releaseMirror.enabled=false` (its false by default so you do not need to set this) | Install, upgrade and uninstall completes successfully|
| Install, Upgrade, and Uninstall with Rancher PSP enabled and set to restricted | Install, upgrade, and Uninstall completes successfully | 
| Install, Upgrade, and Uninstall with Rancher PSP enabled and set to restricted, additionally with `installer.releaseMirror.enabled=true` | Install, upgrade, and Uninstall completes successfully | 
| Rancher PSP enabled and set to restricted: Install with `installer.releaseMirror.enabled=false` (its false by default so you do not need to set this), Upgrade with `installer.releaseMirror.enabled=true`,  Uninstall  with `installer.releaseMirror.enabled=false` (its false by default so you do not need to set this) | Install, Upgrade and Uninstall completes successfully |

**Comment by nickgerace** [comment]
Available on v2.6.1-rc3

**Comment by MSpencer87** [comment]
Unfortunately, I was still able to reproduce this issue on v2.6.1-rc3 when upgrading Istio with installer.releaseMirror.enabled=true

`Upgrade version check passed: 1.9.6 -> 1.10.4.

2021-09-29T20:12:59.053896Z     info    Error: failed to generate Istio configs from file [/app/istio-base.yaml] for the current version: 1.9.6, error: Get "https://github.com/istio/istio/releases/download/1.9.6/istio-1.9.6-linux-amd64.tar.gz": dial tcp 127.0.0.1:443: connect: connection refused

Error: failed to generate Istio configs from file [/app/istio-base.yaml] for the current version: 1.9.6, error: Get "https://github.com/istio/istio/releases/download/1.9.6/istio-1.9.6-linux-amd64.tar.gz": dial tcp 127.0.0.1:443: connect: connection refused
error found during istioctl upgrade`

**Comment by nickgerace** [comment]
Available on `v2.6.1-rc10`

**Comment by MSpencer87** [comment]
Verified Istio upgrade in air-gapped environment

1.9.6 -> 100.0.1+up1.10.4 Successful 

Rancher Version: v2.6.1-rc7
Istio Installer: 1.10.4-rancher3 (added from v2.6.1-rc10)
Provider: rke
Kubernetes Version: rke	v1.21.5

**Comment by brendarearden** [comment]
### My portion of validation: PASSED

|Test Case | Expected Outcome | Status |
|-- | -- | ---|
|Rancher PSP enabled and set to restricted: Install with installer.releaseMirror.enabled=false (its false by default so you do not need to set this), Upgrade with installer.releaseMirror.enabled=true, Uninstall with installer.releaseMirror.enabled=false (its false by default so you do not need to set this) | Install, Upgrade and Uninstall completes successfully | Passed |


### Steps to test:
Scenarios tested in Rancher v2.6.1-rc10, downstream RKE DO cluster running k8s 1.25.5

#### Test Case: **Install, Uninstall with releaseMirror.enabled=false, Upgrade with releaseMirror.enabled=true with Rancher PSP enabled and restricted**

1. Enabled Rancher PSP and set to restricted on downstream cluster
2. Install rancher-istio 100.0.1+up1.10.4 with installer.releaseMirror.enabled=false, all other values left as default
3. Upgrade rancher-istio 100.0.1+up1.10.4 (technically an edit because it was not a version increase) with installer.releaseMirror.enabled=true, all other values left as default
4. Uninstall rancher-istio 100.0.1+up1.10.4

**Comment by rancherdev** [comment]
Verified on v2.6.1-rc10




Test Case | Expected Outcome
-- | --
Install, Upgrade, and Uninstall with Rancher PSP enabled and set to restricted | Install, upgrade, and Uninstall completes successfully

- Create a node driver downstream cluster with PSP set to restricted
- Installed istio version version[v1.10.4] - all workloads are up, kiali and jaeger are accessible - PASS
- Uninstalled istio - all workloads are removed - PASS 
- Installed istio [100.0.0+up1.10.4], upgraded to v1.10.4 - PASS


Test Case | Expected Outcome
-- | --
Install, Upgrade, and Uninstall with Rancher PSP enabled and set to restricted, additionally with installer.releaseMirror.enabled=true | Install, upgrade, and Uninstall completes successfully



- Create a node driver downstream cluster with PSP set to restricted
- Installed istio version version[v1.10.4] - all workloads are up, kiali and jaeger are accessible - PASS
- Update the istio by setting `installer.releaseMirror.enabled=true` in the yaml while installing istio - PASS
- Uninstalled istio - all workloads are removed - PASS 
- Installed istio [100.0.0+up1.10.4], upgraded to v1.10.4 with `installer.releaseMirror.enabled=true` - PASS

**Comment by anupama2501** [comment]
Verified on v2.6.1-rc10




Test Case | Expected Outcome
-- | --
Install, Upgrade, and Uninstall with Rancher PSP enabled and set to restricted | Install, upgrade, and Uninstall completes successfully

- Create a node driver downstream cluster with PSP set to restricted
- Installed istio version version[v1.10.4] - all workloads are up, kiali and jaeger are accessible - PASS
- Uninstalled istio - all workloads are removed - PASS 
- Installed istio [100.0.0+up1.10.4], upgraded to v1.10.4 - PASS


Test Case | Expected Outcome
-- | --
Install, Upgrade, and Uninstall with Rancher PSP enabled and set to restricted, additionally with installer.releaseMirror.enabled=true | Install, upgrade, and Uninstall completes successfully



- Create a node driver downstream cluster with PSP set to restricted
- Installed istio version version[v1.10.4] - all workloads are up, kiali and jaeger are accessible - PASS
- Update the istio by setting `installer.releaseMirror.enabled=true` in the yaml while installing istio - PASS
- Uninstalled istio - all workloads are removed - PASS 
- Installed istio [100.0.0+up1.10.4], upgraded to v1.10.4 with `installer.releaseMirror.enabled=true` - PASS

**Comment by anupama2501** [comment]
Closing this ticket as all the required validations look good.

**Comment by brendarearden** [comment]
Reopened due to a regression that was found. The permission changes that were added to the `istio-installer` have made the `/usr/local/app/dashboards` inaccessible to the nginx user. This causes the datasource not to get set properly on the dashboard config maps. 

#### Steps to reproduce
Env: Rancher v2.6.1-rc10, downstream RKE DO cluster running k8s 1.25.5, rancher-monitoring installed as pre-req for kiali. 
1. Install rancher-istio 100.0.1+up1.10.4 with default settings
2. While install is running, go to workloads>pods>istio-installer and view logs
3. See the following error in the istio-installer logs
```
Installation completesed: can't create temp file '/usr/local/app/dashboards/11829.jsonXXXXXX': Permission denied
--
Thu, Sep 30 2021 11:58:47 am | configmap/istio-dashboard-11829 created
Thu, Sep 30 2021 11:58:47 am | configmap/istio-dashboard-11829 labeled
Thu, Sep 30 2021 11:58:47 am | configmap/istio-dashboard-11829 labeled
Thu, Sep 30 2021 11:58:47 am | sed: can't create temp file '/usr/local/app/dashboards/7630.jsonXXXXXX': Permission denied
Thu, Sep 30 2021 11:58:47 am | configmap/istio-dashboard-7630 created
Thu, Sep 30 2021 11:58:48 am | configmap/istio-dashboard-7630 labeled
Thu, Sep 30 2021 11:58:48 am | configmap/istio-dashboard-7630 labeled
Thu, Sep 30 2021 11:58:48 am | sed: can't create temp file '/usr/local/app/dashboards/7636.jsonXXXXXX': Permission denied
Thu, Sep 30 2021 11:58:48 am | configmap/istio-dashboard-7636 created
Thu, Sep 30 2021 11:58:48 am | configmap/istio-dashboard-7636 labeled
Thu, Sep 30 2021 11:58:49 am | configmap/istio-dashboard-7636 labeled
Thu, Sep 30 2021 11:58:49 am | sed: can't create temp file '/usr/local/app/dashboards/7639.jsonXXXXXX': Permission denied
Thu, Sep 30 2021 11:58:49 am | configmap/istio-dashboard-7639 created
Thu, Sep 30 2021 11:58:49 am | configmap/istio-dashboard-7639 labeled
Thu, Sep 30 2021 11:58:49 am | configmap/istio-dashboard-7639 labeled
Thu, Sep 30 2021 11:58:49 am | sed: can't create temp file '/usr/local/app/dashboards/7642.jsonXXXXXX': Permission denied
Thu, Sep 30 2021 11:58:49 am | configmap/istio-dashboard-7642 created
Thu, Sep 30 2021 11:58:50 am | configmap/istio-dashboard-7642 labeled
Thu, Sep 30 2021 11:58:50 am | configmap/istio-dashboard-7642 labeled
Thu, Sep 30 2021 11:58:50 am | sed: can't create temp file '/usr/local/app/dashboards/7645.jsonXXXXXX': Permission denied
Thu, Sep 30 2021 11:58:50 am | configmap/istio-dashboard-7645 created
Thu, Sep 30 2021 11:58:50 am | configmap/istio-dashboard-7645 labeled
Thu, Sep 30 2021 11:58:50 am | configmap/istio-dashboard-7645 labeled
```
4. After install completes successfully, navigate to storage/configmaps and open one of the `istio-dashboard-<number>` configmaps and see that the `datasource` field is still set to `${DS_PROMETHEUS}` throughout the configmap.

**Comment by anupama2501** [comment]
@brendarearden the helm logs show successful installation of istio and have caught the error but from the helm logs no errors are seen. Do we expect helm to catch these? 
<img width="400" alt="Screen Shot 2021-09-30 at 2 40 10 PM" src="https://user-images.githubusercontent.com/38144301/135534356-dd7bf47e-7bfb-4e3b-97d9-37cdb1d651b6.png">
.

**Comment by brendarearden** [comment]
@anupama2501 The istio-installer pod does not exit the pod on every command error, so for this particular error it is not expected that helm would error.

**Comment by nickgerace** [comment]
Available in v2.6.1-rc11

**Comment by cbron** [comment]
@lion200816 you should be able to fix this problem by setting `installer.debug.secondsSleep` to something on the chart values, that will stall the job. Then exec into the job pod, sudo to root, and then run `echo "hosts: files dns" > /etc/nsswitch.conf`. Lmk how that works for you.
